src/main/java/com/streamr/client/utils/EncryptionUtil.java

```diff
@@ -1,28 +1,42 @@
 package com.streamr.client.utils;
 
 import com.streamr.client.exceptions.InvalidGroupKeyException;
+import com.streamr.client.exceptions.UnableToDecryptException;
+import com.streamr.client.protocol.message_layer.StreamMessage;
 import org.apache.commons.codec.binary.Hex;
 import org.apache.logging.log4j.LogManager;
 import org.apache.logging.log4j.Logger;
 
 import javax.crypto.Cipher;
 import javax.crypto.SecretKey;
 import javax.crypto.spec.IvParameterSpec;
+import javax.crypto.spec.SecretKeySpec;
 import javax.xml.bind.DatatypeConverter;
+import java.io.IOException;
 import java.security.SecureRandom;
+import java.util.Arrays;
 
 public class EncryptionUtil {
     private static final Logger log = LogManager.getLogger();
     private static final SecureRandom SRAND = new SecureRandom();
+    private static final Cipher cipher = getCipher();
```

Comment:
`Cipher` is not thread-safe, so we should either `synchronize` on it, or wrap it in `ThreadLocal` to make `EncryptionUtil` thread-safe again. I recommend the `ThreadLocal` approach because `synchronize` would limit performance for no good reason.